Create a new end point for carrier-types

harriswong opened the pull request:
## Description
This address the first bullet in https://github.com/Automattic/woocommerce-services/issues/2234. 
> Creating an endpoint in the plugin that can be used by the React front-end to get the fields from the Connect Server

### Related issue(s)
https://github.com/Automattic/woocommerce-services/issues/2234

### Steps to reproduce & screenshots/GIFs
_If you can't make a request directly to your own localhost WC API. Refer to the "Steps to test with manual WooCommerce API Client"._
1. Assumes `shipping/carrier-types` end point is available from the server. You can use this [PR](https://github.com/Automattic/woocommerce-connect-server/pull/1597) if needed.
2. Make a GET request to your localhost's WooCommerce Services API `https://localhost/wp-json/wc/v1/connect/shipping/carrier-types` 
3. You should see a a similar response back after applying `json_decode()`:
```
stdClass Object
(
    [carriers] => Array
        (
            [0] => stdClass Object
                (
                    [type] => DhlExpressAccount
                    [name] => DHL Express
                    [fields] => stdClass Object
                        (
                            [account_number] => stdClass Object
                                (
                                    [visibility] => visible
                                    [label] => DHL Account Number
                                )
...
```

### Steps to test with manual WooCommerce API Client
1. We will be using this https://docs.woocommerce.com/document/woocommerce-rest-api/ to make a completely separate script to make API calls.
1. Create a new folder `mkdir woocommerce-api-test`
2. `composer install automattic/woocommerce` ([link](https://packagist.org/packages/automattic/woocommerce))
3. vim `index.php` 
4. paste this in there
```
<?php

require __DIR__ . '/vendor/autoload.php';

use Automattic\WooCommerce\Client;

$woocommerce = new Client(
    'http://localhost',
    'ck_***', // this needs to be your own key
    'cs_***', // this needs to be your own secrets
    [
        'version' => 'wc/v1',
    ]
);

print_r($woocommerce->get('connect/shipping/carrier-types'));
```

5. For the keys and secrets, follow the instructions [here](https://docs.woocommerce.com/document/woocommerce-rest-api/). 
6. run `php index.php` in your console.
7. You should see the results back.
### Checklist

<!-- All testable code should have tests. -->
- [x] unit tests

<!-- An entry in the changelog file is always required -->
- [x] `changelog.txt` entry added

harriswong commented on tests/php/classes/test-class-wc-rest-connect-shipping-carrier-types-controller.php:
```diff
@@ -0,0 +1,86 @@
+		require_once dirname( __FILE__ ) . '/../../../classes/class-wc-connect-service-settings-store.php';
+		require_once dirname( __FILE__ ) . '/../../../classes/class-wc-connect-logger.php';
+		require_once dirname( __FILE__ ) . '/../../../classes/class-wc-rest-connect-base-controller.php';
+		require_once dirname( __FILE__ ) . '/../../../classes/class-wc-rest-connect-shipping-carrier-types-controller.php';
```
These files are imported this way because our unit test bootstrap doesn't autoload all files for the plugin. I think we can improve this either by 
1. Create a test bootstrap for all these files, for all tests in WCS
2. Create an autoloader for WCS to register all WCS scripts.

But probably not this PR, something in the future.

frosso commented on classes/class-wc-rest-connect-shipping-carrier-types-controller.php:
```diff
@@ -0,0 +1,37 @@
+}
+
+if ( class_exists( 'WC_REST_Connect_Shipping_Carrier_Types_Controller' ) ) {
+	return;
```
I love the early return you did here 💯

frosso reviewed (approved):
Looking good so far 👍

harriswong commented:
@frosso please take a look at the last 2 commits. I made the following changes.

- Fixed conflicted changelog
- Updated API response to include a `success` status
- Updated test.

frosso commented on classes/class-wc-rest-connect-shipping-carrier-types-controller.php:
```diff
@@ -0,0 +1,42 @@
+		return new WP_REST_Response(
+			[
+				'success' => true,
+				'fields'  => $response,
```
Something I didn't pay attention to earlier - I am seeing the data returned in `fields > carriers`.

![Screen Shot 2020-11-02 at 3 48 25 PM](https://user-images.githubusercontent.com/273592/97923065-092df600-1d23-11eb-8739-1d5ef7d298ea.png)


Was there a reason why why having this structure? Should it just be `carriers => []`, rather than `fields => carriers => []`?

i.e.:
```suggestion
				'carriers'  => $response->carriers,
```
![Screen Shot 2020-11-02 at 3 52 05 PM](https://user-images.githubusercontent.com/273592/97923310-65911580-1d23-11eb-82a6-5c00603b2c66.png)

harriswong replied:
> Was there a reason why why having this structure? 

I was thinking more of `success => true, data => {...}`. But `data` was too generic, so i chose `fields`. I like the suggestion. We don't need another layer to wrap the `carriers` object.

frosso replied:
Ah gotcha, that makes sense! Yeah, either `data => []` or `carriers => []` work for me! 👍

harriswong replied:
Thanks, updated!

frosso replied:
Looks great, thanks for considering it!
![Screen Shot 2020-11-02 at 4 27 25 PM](https://user-images.githubusercontent.com/273592/97926177-552f6980-1d28-11eb-80a8-e14262cfe84d.png)

frosso reviewed (approved):
Still looks good, thanks for checking! Just some considerations on the response being returned.
